Pickle CyFunctions

da-woods opened the pull request:
Support pickling of CyFunctions.

Will eventually close #2584

Follows the scheme outlines in the comments of that issue. I'm using the function cname as the key (because I believe it should be unique).

Currently known to not be working are:
* fused functions (except for a few select very trivial cases)
* any function with a classobj - although it isn't easy to generate these in a closure in a way that's ever likely to be useful.
* any function with non-trivial defaults - this requires a fairly substantial change to how defaults are stored, which I think isn't needed for the initial implementation.
* Running generators and any generator expressions (would need a separate type of be changed)

All of these cases fail fairly gracefully so I'm not planning to deal with them here.

da-woods committed:
Expanded reduce function for cyfunction

It still requires a bunch of initialization of the Cython side

da-woods committed:
Create function pointer lookup func

da-woods committed:
Make closure classes pickleable if possible

da-woods committed:
Can unpickle the closures now

da-woods committed:
Closured class optionally initializes Python members

da-woods committed:
Basic prototype works

da-woods committed:
Made and passed basic tests

da-woods commented:
And the immediate conclusion is that trying to pickle anything with a defaults tuple fails...

0dminnimda commented:
FWIW
```python
def foo(kw=[]):
    print(kw)
    kw.append(0)

foo()  # []
foo()  # [0]

# pickle foo
bar = ...  # unpickle

bar()  # [0, 0]
```

We may want to pickle the current state of the default value and not just the initial default, that is set while creating the function object (if the same thing happens in the CPython).

da-woods commented:
> (if the same thing happens in the CPython).

CPython only pickles functions by storing their qualified name. So I'm pretty sure it loses any modifiable state. (It pickles methods as `self`+the function though). So it we can't use CPython as a reference here.

da-woods commented:
Looking at it some more, The way Cython handles defaults is that the defaults tuple and default keyword arg dict is purely for introspection and that Cython actually uses a `void*` to an opaque defaults structure for the "actual" defaults. I think that means that most things with default arguments are inherently unpickleable with the current design.

da-woods commented:
I think the long-term solution is to change `defaults` from an opaque struct to a custom Python (much like the closure classes). That would probably simplify other bits of the implementation too, at the cost of a little bit more memory used for simple cases

The short-term solution is probably just to ban pickling for any non-Python default argument.

Either way I'll close this for now since it's probably further from working that I thought.

da-woods committed:
Ban pickling of difficult cases

da-woods committed:
More tests

da-woods committed:
Can handle defauls for introspection

da-woods committed:
Lambdas work

da-woods committed:
Excluded genexprs somewhat

They don't work yet and they cause issues.
Also hid internal scope

da-woods committed:
Fused functions works as well as they can for now

da-woods committed:
Avoid creating a new FusedFunction on class lookup

Among other things this makes it pickleable by ensuring
that it's the same object each time

da-woods committed:
Hopefully fixed most of the tests

da-woods committed:
Changed representation of is_internal

(None was a little awkward to work with)

da-woods committed:
Handle pickleable funcs in utility code

da-woods committed:
Fixed "all" build

da-woods committed:
Fixed bug with unpickleable closures

da-woods committed:
Fix unicode names issue

da-woods committed:
More testsuite bugfixes

One missing attribute on compiled ParseTreeTransforms
Exclude Python 3.4 from some tests
Exclude limited API

0dminnimda commented on Cython/Compiler/ModuleNode.py:
```diff
@@ -1592,13 +1593,23 @@ def generate_new_function(self, scope, code, cclass_entry):
             code.putln("new((void*)&(p->%s)) %s();" % (
                 entry.cname, decl_code))
 
+        if scope.is_internal == 2 and py_attrs:
```
I have a feeling that `is_x` should only be `True` or `False` and if we want to represent more states then it should not be `is_x`

da-woods replied:
Yes that's probably right. I'll replace it with a better name at the next round of revisions (once I've thought of a better name..)

da-woods commented:
I think this mostly works now. The remaining issues are related to https://github.com/cython/cython/issues/4373 with limited api and type-specs. I can't see an easy way to add a bugs file for type-specs right now.

There's obviously some limitations about what it can do, but I think addressing those would make it a much larger PR so it's better left for later.

da-woods committed:
Fix fused-types module

da-woods committed:
Updated limited_api_bugs

da-woods committed:
Changed name of "is_internal"

0dminnimda commented on Cython/Compiler/ModuleNode.py:
```diff
@@ -1592,13 +1593,23 @@ def generate_new_function(self, scope, code, cclass_entry):
                     entry.cname, entry.cname))
             else:
                 code.put_init_var_to_py_none(entry, "p->%s", nanny=False)
+        if scope.internal_mode == 2 and py_attrs:
+            code.putln("}")
```
I think it would be better to use enum (or at least strings) to indicate the state

0dminnimda replied:
enum requires a little more text and imports but is less error-prone

Oh yeah .. there is no enum in python 2.7 ... but we can use the enum34 backport!

da-woods replied:
> but we can use the enum34 backport!

No. I think we try quite hard not to require extra dependencies. So while I like enums it isn't worth adding a dependency for that.

0dminnimda replied:
> I think we try quite hard not to require extra dependencies.

Is there any reason for this? If our code is cleaner and limits unnecessary freedom, then the quality of the product itself will improve.

Not to mention the fact that we can probably refuse to support running cython itself on python 2.7, while still allowing compilation for this version.

da-woods replied:
This PR isn't really the place to debate this so I'll reply once and then leave the topic, but:

> Is there any reason for this? If our code is cleaner and limits unnecessary freedom, then the quality of the product itself will improve.

Adding dependencies would Cython harder to install (you can no longer tell users "just install Cython"...), you have to track versions for dependencies (have they changed and broken anything, is it a problem if users have another a different version installed?), and you can't necessarily control the dependencies they bring in. Of course it's a trade-off and there's definitely advantages to using good code written elsewhere, but for something like Cython that's a dependency of a lot of other stuff there's a big advantage to having very minimal decisions. (Of course, this wasn't my decision when it was originally made, and I can't know all the motivations of the people who made it...)

> Not to mention the fact that we can probably refuse to support running cython itself on python 2.7, while still allowing compilation for this version.

That's probably as good as dropping support entirely: if I want a Cython module compiling with a specific version of Python I call `pythonx.x setup.py ...`, or `cythonize-x.x` (which uses that version of Python), or use a Jupyter notebook with that version of Python. Of course it's possible to do the steps separately but it's hard and most people don't.

0dminnimda replied:
> This PR isn't really the place to debate this

True, but where is that place?

> Of course it's possible to do the steps separately but it's hard and most people don't.

I think it's the matter of giving the users such mechanism.
Like with the `ast.parse` (`feature_version` argument), we can just apply different rules and restrictions depending on the versions for what we compile.
I don't know how long it will take to end support for python 2, it seems to me long enough to create an artificial support mechanism for compiling for earlier versions.

There may be some caveats with compiling for python 2 with the library of python 3, but we can either request the presence of the python 2 files, or (the better but longer/harder way) just simulate or copypaste some implementation details, that are not present in python 3

scoder replied:
> > This PR isn't really the place to debate this
> 
> True, but where is that place?

The cython-devel mailing list. But please don't ask for this specific addition as I would only reject it.

> Not to mention the fact that we can probably refuse to support running cython itself on python 2.7, while still allowing compilation for this version.

This would make Cython substantially more difficult to install and test. Also not up for discussion right now. Py2.7 support will be removed in Cython 3.1, so there is no need to even think about this.

0dminnimda replied:
> But please don't ask for this specific addition as I would only reject it.

I don't consider this to be a reason, please, don't think about me like that.

da-woods committed:
Simpler type-spec workaround for __module__

da-woods committed:
Small suggestions from review

da-woods commented on Cython/Compiler/Nodes.py:
```diff
@@ -5486,8 +5486,8 @@ def generate_type_ready_code(entry, code, bases_tuple_cname=None, check_heap_typ
+            if not type.scope.internal_mode and not type.scope.directives.get('internal'):
+                # scope.internal_mode is set for types defined by
                 # Cython (such as closures), the 'internal'
                 # directive is set by users
```
I've made a small change to the wording here

scoder commented on Cython/Compiler/ModuleNode.py:
```diff
@@ -1592,13 +1593,23 @@ def generate_new_function(self, scope, code, cclass_entry):
             code.putln("new((void*)&(p->%s)) %s();" % (
                 entry.cname, decl_code))
 
+        if scope.internal_mode == 2 and py_attrs:
+            # create a hybrid "optional initialization" where the kwds
+            # argument is used to signal that initialization happens.
+            # This is needed for pickleable closures, where the attributes
```
Since this is only "for pickleable closures", there's probably a better way to indicate this specific case than abusing the user controllable `is_internal` option (i.e. `@cython.internal`).

da-woods replied:
I've replaced it with a `needs_pickleable_closure_constructor` flag and reverted `is_internal` back to its original use. Hopefully that's clear, if verbose.

(I think the `is_internal` flag always did more than the user `@cython.internal` option though and is only used for closure scopes)

da-woods committed:
Added "needs_pickleable_closure_constructor" flag

da-woods committed:
Merge branch 'master' into pickle_functions

da-woods committed:
Merge branch 'master' into pickle_functions

da-woods committed:
Move lookup to dict and switch

Instead of many string comparisons. I think this'll be a bit more
optimized

da-woods committed:
Merge remote-tracking branch 'origin/pickle_functions' into pickle_functions

da-woods committed:
Style

da-woods committed:
Merge in pickle-cnames too

da-woods committed:
Handle platforms where sizeof(funcptr) != sizeof(void*)

Fixes #4683

It's mostly a theoretical concern I think, but may apply to
some "not too obscure" platforms (ia64?).

For the majority of platforms it keeps working as before (copies
the function pointer into the void* of a capsule) although I've
used memcpy rather than a union to do this (because I don't
believe the union trick is 100% guaranteed to work in standard
c++, although practically it does of course).

For any where this doesn't apply it allocates memory instead and
the capsule points to that.

My interest in this is to use when pickling CyFunctions, but this
PR just uses the mechanism for cimporting cdef functions

da-woods committed:
Convert funcptr->pyobject using shared mechanism

da-woods committed:
Undid changes to "ImportExport"

they're really part of a separate PR

da-woods committed:
...and the associated test

da-woods committed:
Merge branch 'master' into pickle_functions

da-woods committed:
Merge branch 'master' into pickle_functions

da-woods committed:
Fix issue with genexp dummpy names being pickled

da-woods committed:
Fix codestyle error

[skip ci]

scoder committed:
Merge branch 'master' into pickle_functions

da-woods committed:
Let the functionpickling utility code come at the end

Since it needs access to all the C function pointers defined in
earlier Cython code.

da-woods committed:
Remove unused variable

da-woods committed:
Update a few comments

da-woods committed:
Merge branch 'master' into pickle_functions

da-woods committed:
Make it compile on the limited api

da-woods committed:
Merge branch 'master' into pickle_functions

da-woods committed:
Update tests

da-woods committed:
Move some code into Nodes.py

da-woods committed:
Fix setup.py

da-woods committed:
Fix annotation-related cleanup

da-woods committed:
code style

yisun-anetac commented:
@da-woods is there any plan to continue the development for this PR? It is a very important feature we are looking for, really hope there is a solution.
Thanks
Yi

da-woods committed:
Merge remote-tracking branch 'real_origin/master' into pickle_functions

da-woods committed:
Simplify void* handling

Since I now believe they have to be the same size as void* or
Python breaks.

da-woods committed:
Fix build

da-woods committed:
Fixes and small optimization

da-woods committed:
Label at end error

da-woods committed:
Fix warning

da-woods committed:
Fix tests, remove stray print

da-woods committed:
Fix Limited API

da-woods committed:
Version-specific exclude for limited-api tests

da-woods committed:
Merge remote-tracking branch 'real_origin/master' into pickle_functions

da-woods committed:
Fix Limited API and simplify globals

da-woods committed:
Binary search for name instead
